File: Array/Find the Number of Distinct Colors Among the Balls/Find_the_Number_of_Distinct_Colors_Among_the_Balls.java
// Problem Name: Find the Number of Distinct Colors Among the Balls
// Problem Link: https://leetcode.com/problems/find-the-number-of-distinct-colors-among-the-balls/description/

class Solution {
    public int[] queryResults(int limit, int[][] queries) {
        int n = queries.length;
        int[] ans = new int[n];
        int k = 0;
        Map<Integer, Integer> mp1 = new HashMap<>();
        Map<Integer, Integer> mp2 = new HashMap<>();
        for(int[] q : queries) {
            int ball = q[0];
            int color = q[1];
            if(!mp1.containsKey(ball)) {
                mp1.put(ball, color);
                mp2.put(color, mp2.getOrDefault(color, 0) + 1);
            } else {
                if(mp1.get(ball) != color) {
                    int prevColor = mp1.get(ball);
                    mp2.put(prevColor, mp2.get(prevColor) - 1);
                    if (mp2.get(prevColor) == 0) {
                        mp2.remove(prevColor);
                    }
                    mp1.put(ball, color);
                    mp2.put(color, mp2.getOrDefault(color, 0) + 1);
                }
            }
            ans[k++] = mp2.size();
        }
        return ans;
    }
}

File: Array/Find the Number of Distinct Colors Among the Balls/Find_the_Number_of_Distinct_Colors_Among_the_Balls.cpp
// Problem Name: Find the Number of Distinct Colors Among the Balls
// Problem Link: https://leetcode.com/problems/find-the-number-of-distinct-colors-among-the-balls/description/

class Solution {
public:
    vector<int> queryResults(int limit, vector<vector<int>>& queries) {
        vector<int> ans;
        unordered_map<int, int> mp1, mp2;
        for (auto& q : queries) {
            int ball = q[0];
            int color = q[1];
            if (mp1.find(ball) == mp1.end()) {
                mp2[color]++;
                mp1[ball] = color;
            } else {
                if (mp1[ball] != color) {
                    mp2[mp1[ball]]--;
                    if (mp2[mp1[ball]] == 0) {
                        mp2.erase(mp1[ball]);
                    }
                    mp1[ball] = color;
                    mp2[color]++;
                }
            }
            ans.push_back(mp2.size());
        }
        return ans;
    }
};